rapids-cmake/cpm/generate_pinned_versions.cmake

```diff
@@ -0,0 +1,107 @@
+#=============================================================================
+# Copyright (c) 2024, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+include_guard(GLOBAL)
+
+#[=======================================================================[.rst:
+rapids_cpm_generate_pinned_versions
+-----------------------------------
+
+.. versionadded:: v24.04.00
+
+Generate a json file with all dependencies with pinned version values
+
+.. code-block:: cmake
+
+  rapids_cpm_generate_pinned_versions( OUTPUT <json_verions_output_path> )
+
+Generates a json file with all cpm dependencies with pinned version values.
+This allows for subsequent reproducible builds using the exact same state.
+
+The rapids-cmake default `versions.json` uses branch names or git tag names
+for dependencies. This is done so that projects can 'live at head' of dependencies.
+By using :cmake:command:`rapids_cpm_package_override` a project can specify a custom
+`versions.json` that specifies exact git SHA's so that projects have reproducible builds.
+
+:cmake:command:`rapids_cpm_generate_pinned_versions` can be used to transform a set of
+rapids-cmake dependencies from branch names to pinned values. This can be used in subsequent
+builds, e.g:
+
+  1. Have CI run with `versions.json` which tracks dependency by branch name
+  2. Store the generated pinned `versions.json` from the CI builds
+  3. If build is good, create the release branch and commit the generated pinned `versions.json`
+     to have reproducible builds for that release
+
+``OUTPUT``
+Specify a file path to write the pinned versions information to. Can be called multiple
+times and each unique path will be written too.
+
+The generated json file will have the following entries for each package:
+
+.. code-block:: json
+
+  {
+    "version" : "<CPM_PACKAGE_<package_name>_VERSION>",
+    "git_url" : "<deduced>",
+    "git_tag" : "<deduced>",
+    "git_shallow" : "false",
+    "always_download" : "true"
+  }
+
+
+If the original package ( or override ) also had any `patches`, or `proprietary_binary`
+fields those will be propagated to the generated entry.
+
+.. note::
+  The git SHA1 computed for each package is found be finding the most recent
+  commit that can be cloned from the url.
```

Comment:
I'm not sure if this is a good design. I think the better behavior would be to write out a pinning with the actual commit used, regardless of whether it exists on the remote. Pushing commits to the remote to ensure reproducibility is a reasonable responsibility for the user and would simplify the implementation of this feature. This might also be important to support multiple remotes, where a commit might only exist on a private fork or user's fork but not the upstream URL. Perhaps we can emit a warning if that SHA doesn't exist on the remote.

Reply:
We can't use the latest git commit. If the project has patches with the `.patch` suffix they are applied as git commits ( `git am` ). Therefore these patches will move `HEAD` and therefore that will not be clone-able when trying to reproduce the project.

> This might also be important to support multiple remotes, where a commit might only exist on a private fork or user's fork but not the upstream URL. Perhaps we can emit a warning if that SHA doesn't exist on the remote.

When using rapids-cmake, cpm, fetch_content, etc from CMake you only have one remote in a git cloned project. But even in the case the local user has changed that the `rapids_cpm_gpv_extract_source_git_info ` extracts the remote details for the current branch via the 'tracking' details

Reply:
Okay. I'm not sure if I entirely understand this, but I trust your reasoning.